pkg/pillar/cmd/zedagent/localinfo.go

```diff
@@ -192,73 +195,133 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
 			continue
 		}
 		// Try to find the application instance.
-		ais := findAppInstance(ctx, appUUID, displayName)
-		if ais == nil {
+		appInst := findAppInstance(ctx, appUUID, displayName)
+		if appInst == nil {
 			log.Warnf("Failed to find app instance with UUID=%s, displayName=%s",
 				appUUID, displayName)
 			continue
 		}
-		appUUID = ais.UUIDandVersion.UUID
+		appUUID = appInst.UUIDandVersion.UUID
+		if _, duplicate := publishApps[appUUID.String()]; duplicate {
+			log.Warnf("Multiple commands requested for app instance with UUID=%s",
+				appUUID)
+			continue
+		}
+
 		command := types.AppCommand(appCmdReq.Command)
-		appCmd := ctx.localAppCommands.LookupByAppUUID(appUUID)
-		if appCmd != nil {
-			// Entry for this app already exists.
-			if appCmd.Command == command &&
-				appCmd.LocalServerTimestamp == appCmdReq.Timestamp {
-				// already accepted
-				continue
+		appCmd := ctx.localAppCommands[appUUID.String()]
+		if appCmd == nil {
+			appCmd = &types.LocalAppCommand{
+				AppUUID:          appUUID,
+				VolumeGenCounter: make(map[string]int64),
 			}
-			appCmd.Command = command
-			appCmd.LocalServerTimestamp = appCmdReq.Timestamp
-			appCmd.DeviceTimestamp = time.Now()
-			appCmd.Completed = false
-			cmdChanges = true
+			ctx.localAppCommands[appUUID.String()] = appCmd
+		}
+		if appCmd.Command == command &&
+			appCmd.LocalServerTimestamp == appCmdReq.Timestamp {
+			// already accepted
 			continue
 		}
-		// Add new entry.
-		ctx.localAppCommands.Cmds = append(ctx.localAppCommands.Cmds,
-			types.LocalAppCommand{
-				AppUUID:              appUUID,
-				Command:              command,
-				LocalServerTimestamp: appCmdReq.Timestamp,
-				DeviceTimestamp:      time.Now(),
-				Completed:            false,
-			})
+		appCmd.Command = command
+		appCmd.LocalServerTimestamp = appCmdReq.Timestamp
+		appCmd.DeviceTimestamp = time.Now()
+		appCmd.Completed = false
+
+		// Update configuration to trigger the operation.
+		switch appCmd.Command {
+		case types.AppCommandRestart:
+			appCmd.RestartCounter++
+			appInst.RestartCmd.Counter++
+			publishApps[appUUID.String()] = *appInst
+
+		case types.AppCommandPurge:
+			appCmd.PurgeCounter++
+			appInst.PurgeCmd.Counter++
+			for i := range appInst.VolumeRefConfigList {
+				vr := &appInst.VolumeRefConfigList[i]
+				volKey := volumeKey(vr.VolumeID.String(), vr.GenerationCounter)
+				volObj, _ := ctx.pubVolumeConfig.Get(volKey)
+				if volObj == nil {
+					log.Warnf("Failed to find volume %s referenced by app instance "+
+						"with UUID=%s - not purging this volume", volKey, appUUID)
+					continue
+				}
+				unpublishVols[volKey] = struct{}{}
+				vr.GenerationCounter++
+				volume := volObj.(types.VolumeConfig)
+				volume.GenerationCounter++
+				publishVols[volume.Key()] = volume
+				appCmd.VolumeGenCounter[vr.VolumeID.String()]++
```

Comment:
Should we have configurable option in api to determine which volumes we should keep not-purgeable during this operation?

Reply:
I think we need to add this; the commercial controller already has such a notion internally.
But we need to check what semantics the users want for the local purge - whether they want to purge all the volumes or only the ones which the controller has marked as "purgeable" in some API addition.

Reply:
How about adding documentation to the API document to state that the purge command will purge all the volumes. That way we don't forget.

Reply:
Added comment about this to `local_profile.proto`.